kernel/parray.ml

```diff
@@ -0,0 +1,156 @@
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+let max_array_length32 = 4194303
+
+let max_length = Uint63.of_int max_array_length32
+
+let length_to_int i = snd (Uint63.to_int2 i)
+
+let trunc_size n =
+  if Uint63.le Uint63.zero n && Uint63.lt n (Uint63.of_int max_array_length32) then
+    length_to_int n
+  else max_array_length32
+
+type 'a t = ('a kind) ref
+and 'a kind =
+  | Array of 'a array * 'a
+  | Updated of int * 'a * 'a t
+
+let unsafe_of_array t def = ref (Array (t,def))
+let of_array t def = unsafe_of_array (Array.copy t) def
+
+let rec length p =
+  match !p with
+  | Array (t,_) -> Uint63.of_int @@ Array.length t
+  | Updated (_, _, p) -> length p
+
+let length p =
+  match !p with
+  | Array (t,_) -> Uint63.of_int @@ Array.length t
+  | Updated (_, _, p) -> length p
+
+let rec get_updated p n =
+  match !p with
+  | Array (t,def) ->
+      let l = Array.length t in
+      if Uint63.le Uint63.zero n && Uint63.lt n (Uint63.of_int l) then
+        Array.unsafe_get t (length_to_int n)
+      else def
+  | Updated (k,e,p) ->
+     if Uint63.equal n (Uint63.of_int k) then e
+     else get_updated p n
+
+let get p n =
+  match !p with
+  | Array (t,def) ->
+      let l = Array.length t in
+      if Uint63.le Uint63.zero n && Uint63.lt n (Uint63.of_int l) then
+        Array.unsafe_get t (length_to_int n)
+      else def
+  | Updated _ -> get_updated p n
+
+let set p n e =
+  let kind = !p in
+  match kind with
+  | Array (t,_) ->
+      let l = Uint63.of_int @@ Array.length t in
+      if Uint63.le Uint63.zero n && Uint63.lt n l then
+        let res = ref kind in
+        let n = length_to_int n in
+        p := Updated (n, Array.unsafe_get t n, res);
+        Array.unsafe_set t n e;
+        res
+      else p
+  | Updated _ ->
+      if Uint63.le Uint63.zero n && Uint63.lt n (length p) then
+        ref (Updated((length_to_int n), e, p))
+      else p
+
+let destr_set p n e =
+  let kind = !p in
+  match kind with
+  | Array (t,_) ->
+      let l = Uint63.of_int @@ Array.length t in
+      if Uint63.le Uint63.zero n && Uint63.lt n l then
+        let n = length_to_int n in
+        Array.unsafe_set t n e;
+        p
+      else p
+  | Updated _ -> set p n e
+
+let rec default_updated p =
+  match !p with
+  | Array (_,def) -> def
+  | Updated (_,_,p) -> default_updated p
+
+let default p =
+  match !p with
+  | Array (_,def) -> def
+  | Updated (_,_,p) -> default_updated p
```

Comment:
Why isn't this just 1 function?

Reply:
I don't know, I've been wonderring. These functions are injected as VM values, so I'll confirm with Benjamin whether this is done on purpose or not (to avoid embedding a recursive closure, for some reason). At the very least, it should be commented.

Reply:
Now fixed.